ml-prims/src/hmm/magma/b_likelihood.h

```diff
@@ -0,0 +1,242 @@
+                                idxO = IDX(i, j, lddO);
+                                idxX = IDX(i, j, lddx);
+                                idxY = IDX(i, 0, lddy);
+                                // printf("%f\n", (float) dX_array[bId][idxX] - dY_array[bId][idxY]);
+                                // printf("%d\n", idxO);
+                                dO_array[bId][idxO] = dX_array[bId][idxX] - dY_array[bId][idxY];
+                        }
+                }
+        }
+}
+
+template <typename T>
+void subtract_batched(magma_int_t m, magma_int_t n, magma_int_t batchCount,
+                      T** dO_array, magma_int_t lddO,
+                      T** dX_array, magma_int_t lddx,
+                      T** dY_array, magma_int_t lddy){
+        dim3 block(32, 32, 1);
+        dim3 grid(ceildiv((int)m, (int)block.x),
+                  ceildiv((int)n, (int)block.y),
+                  1);
+
+        int nThreads_x = grid.x * block.x;
+        int nThreads_y = grid.y * block.y;
+        int nThreads_z = grid.z * block.z;
+
+        subtractBatchedKernel<T> <<< grid, block >>>(m, n, batchCount,
+                                                     dO_array, lddO,
+                                                     dX_array, lddx,
+                                                     dY_array, lddy,
+                                                     nThreads_x, nThreads_y, nThreads_z);
+        cudaDeviceSynchronize();
+        CUDA_CHECK(cudaPeekAtLastError());
+}
+
+template <typename T>
+__host__ __device__
+T lol_llhd_atomic(T inv_det, T bil, int nDim){
+        return -0.5 * (-std::log(inv_det) + nDim * std::log(2 * M_PI) + bil);
+}
+
+template <typename T>
+__global__
+void LogLikelihoodKernel(int nObs, int nCl, int nDim,
+                         T* dInvdet_array, T* dBil_batches,
+                         T* dLlhd, int lddLlhd, bool isLog,
+                         int nThreads_x, int nThreads_y){
+        int i_start = threadIdx.x + blockDim.x * blockIdx.x;
+        int j_start = threadIdx.y + blockDim.y * blockIdx.y;
+        int idxL, idxB;
+        for (size_t clId = i_start; clId < nCl; clId+=nThreads_x) {
+                for (size_t oId = j_start; oId < nObs; oId+=nThreads_y) {
+                        idxL = IDX(clId, oId, lddLlhd);
+                        idxB = IDX(clId, oId, nCl);
+
+                        dLlhd[idxL] = lol_llhd_atomic(dInvdet_array[clId],
+                                                      dBil_batches[idxB],
+                                                      nDim);
+
+                        if (!isLog) {
+                                dLlhd[idxL] = std::exp(dLlhd[idxL]);
+                        }
+                }
+        }
+}
+
+
+template <typename T>
+void _likelihood_batched(int nObs, int nCl, int nDim,
+                         T* dInvdet_array, T* dBil_batches,
+                         T* dLlhd, int lddLlhd, bool isLog){
```

Comment:
accept stream argument

Reply:
Similarly it's relevant for the GMM PR, I have deleted it from this PR